Postgres

juancarlospaco opened the pull request:
**Feel free to Edit, Suggest, Comment, Push, etc.**

ThomasTJdev commented:
Really interesting @juancarlospaco!! I'm looking forward to checkout it out - I'm having time in week 3.

1) Is Araq's [ormin](https://github.com/Araq/ormin) production ready?

2) Are you planning to remove sqlite? SQLite is the easy deployment solution for new users, so I would prefer, if we could keep if. We can just use a `-d:sqlite` param for choosing it at compile time.

juancarlospaco commented:
1) Yes.
2) I think with Docker is really easy to have Postgres everywhere, the idea is to provide a ready-to-run Docker too, I am focusing Postgres first.

juancarlospaco commented:
Yes, but I am doing the postgres first on the code, then when it works will do another pass adding a `when`, just to speed up coding, exactly like your code sample.

ThomasTJdev commented on README.md:
```diff
@@ -5,34 +5,50 @@ A quick website tool. Run the nim file and access your webpage. Website: [https:
+- Plugin manager to install, enable and disable plugins directly from browser.
+- Custom title, meta description and keywords for each page, SEO friendly.
+- Custom head, navbar and footer, no hardcoded watermarks, links or logos.
+- Edit core or custom JS and CSS directly from browser, UI/UX Designer friendly.
```
Much better with the descriptive texts!
:+1:

ThomasTJdev commented on nimwcpkg/nimwc_main.nim:
```diff
@@ -182,35 +180,27 @@ macro extensionJs*(): string =
+    echo "Plugins - JS:\n" & result
 
 
+# Loading config file #########################################################
```
See comment about macro ####

ThomasTJdev commented on nimwcpkg/nimwc_main.nim:
```diff
@@ -182,35 +180,27 @@ macro extensionJs*(): string =
-      Loading config file
-__________________________________________________]#
-var db: DbConn
+var db {.global.}: DbConn  # ORMin needs DbConn be var global named "db".
```
Make text documentation ready with double ##?
```nim
var db {.global.}: DbConn  ## ORMin needs DbConn be var global named "db".
```

ThomasTJdev replied:
Maybe not! db initiated in many modules. I could get messy.

ThomasTJdev commented on nimwcpkg/nimwc_main.nim:
```diff
@@ -249,67 +238,76 @@ proc init(c: var TData) =
   c.loggedIn = false
 
 
+# Recompile ###################################################################
```
See comment about macro

ThomasTJdev commented on nimwcpkg/nimwc_main.nim:
```diff
@@ -249,67 +238,76 @@ proc init(c: var TData) =
   return execCmd("nim c " & checkCompileOptions & " -o:nimwcpkg/nimwc_main_new " & getAppDir() & "/nimwc_main.nim")
 
 
+# Validation check ############################################################
```
See comment about macro

ThomasTJdev commented on nimwcpkg/nimwc_main.nim:
```diff
@@ -249,67 +238,76 @@ proc init(c: var TData) =
   c.username.len > 0
 
 
+# Check if user is signed in ##################################################
```
See comment about macro

ThomasTJdev commented on nimwcpkg/nimwc_main.nim:
```diff
@@ -52,6 +55,10 @@ const
 
+  sql_now =
+    when defined(sqlite): "(STRFTIME('%s', 'now'))" # SQLite 3 now() function.
+    else:                 "NOW()"                   # Postgres now() function.
```
If we change from epoch() it would break current logged in users. They can of course just log in again...

ThomasTJdev commented on nimwcpkg/nimwc_main.nim:
```diff
@@ -249,67 +238,76 @@ proc init(c: var TData) =
     c.loggedIn = false
 
 
+# User login ##################################################################
```
See comment about macro

ThomasTJdev commented on nimwcpkg/nimwc_main.nim:
```diff
@@ -249,67 +238,76 @@ proc init(c: var TData) =
-      return (false, "Email may not be test@test.com")
-
-  const query = sql"SELECT id, name, password, email, salt, status, secretUrl FROM person WHERE email = ? AND status <> 'Deactivated'"
+      return (false, "Email must not be 'test@test.com', because is Demo-only.")
```
Do we need to inform the user about the `-d:demo`? It could be a "regular" user without any knowledge about NimWC and what "Demo-only" indicates, then it would only confuse them.
```nim
return (false, "Email must not be 'test@test.com'.")
```

ThomasTJdev commented on nimwcpkg/nimwc_main.nim:
```diff
@@ -249,67 +238,76 @@ proc init(c: var TData) =
+    select person(id, name, password, email, salt, status, secretUrl)
+    where email == ?email.toLowerAscii and status != ?"Deactivated"
+
+  for row in somerows:
```
I like the `let` to indicate what it contains. This only only return 1 row, not rows. What about:
```nim
let userdata =
```

ThomasTJdev commented on nimwcpkg/nimwc_main.nim:
```diff
@@ -335,24 +334,20 @@ proc login(c: var TData, email, pass: string): tuple[b: bool, s: string] =
+    where ip == ?c.req.ip and key == ?c.req.cookies["sid"]
 
 
+# Check if logged in ##########################################################
```
See comment about macros

ThomasTJdev commented on nimwcpkg/nimwc_main.nim:
```diff
@@ -362,49 +357,32 @@ template createTFD() =
   c.loggedIn = loggedIn(c)
 
 
+# HTML tools ##################################################################
```
See comment about macros

ThomasTJdev commented on nimwcpkg/nimwc_main.nim:
```diff
@@ -415,46 +393,35 @@ template statusIntToCheckbox(status, value: string): string =
     ""
 
 
+# Proc's for demo usage #######################################################
```
See comment about macros

ThomasTJdev commented on nimwcpkg/nimwc_main.nim:
```diff
@@ -415,46 +393,35 @@ template statusIntToCheckbox(status, value: string): string =
     ##
-    ## The database will be overwritten with the standard
-    ## standard data every hour.
+    ## The database will be overwritten with standard standard data every hour.
```
Only 1 `standard` :smiley:

ThomasTJdev commented on nimwcpkg/nimwc_main.nim:
```diff
@@ -415,46 +393,35 @@ template statusIntToCheckbox(status, value: string): string =
-    ## This proc is used, when the platform needs to run
-    ## as a test with e.g. public access.
-
+    ## This proc is used when the platform needs to run as a test (public access)
```
Maybe?
```nim
## This proc is used when the platform needs to run as a test or in demo-mode with public access
```

ThomasTJdev commented on nimwcpkg/nimwc_main.nim:
```diff
@@ -415,46 +393,35 @@ template statusIntToCheckbox(status, value: string): string =
     createStandardData(db)
 
 
+# Main module #################################################################
```
See comment about macros

ThomasTJdev commented on nimwcpkg/nimwc_main.nim:
```diff
@@ -512,25 +471,21 @@ when isMainModule:
   info("Up and running!.")
 
 
+# Include HTML files ##########################################################
```
See comment about macros

ThomasTJdev commented on nimwcpkg/nimwc_main.nim:
```diff
@@ -547,16 +502,16 @@ include "tmpl/sitemap.tmpl"
-      Routes for WWW
-__________________________________________________]#
+
+# Routes for WWW ##############################################################
```
See comment about macros

ThomasTJdev commented on nimwcpkg/resources/administration/create_adminuser.nim:
```diff
@@ -1,60 +1,66 @@
-  ## Create new admin user
-  ## Input is done through stdin
-
+  ## Create new admin user. Input is done through stdin.
```
We are not using stdin anymore. Params is passed as args [1].
```nim
## Create new admin user.
```

[1]
https://github.com/ThomasTJdev/nim_websitecreator/pull/39/files#diff-38278c35d32819bc3ba18ef0b3259605R25

ThomasTJdev commented on nimwcpkg/resources/administration/createdb.nim:
```diff
@@ -3,179 +3,179 @@
 #        Hubanize / Cx Manager
```
`Hubanize / Cx Manager` should be removed. I know.. it's a legacy inserted by me..

ThomasTJdev commented on nimwcpkg/resources/administration/createdb.nim:
```diff
@@ -3,179 +3,179 @@
-  TEmail = "varchar(30)"
+  sql_now =
+    when defined(sqlite): "(STRFTIME('%s', 'now'))" # SQLite 3 now() function.
+    else:                 "NOW()"                   # Postgres now() function.
```
Same comment regarding epoch date.

ThomasTJdev commented on nimwcpkg/resources/administration/createdb.nim:
```diff
@@ -3,179 +3,179 @@
+      id         integer       primary key,
+      name       varchar(60)   not null,
+      password   varchar(300)  not null,
+      email      varchar(60)   not null,
```
emailType = "varchar(254)" # https://stackoverflow.com/a/574698/492186

ThomasTJdev commented on nimwcpkg/resources/administration/createdb.nim:
```diff
@@ -3,179 +3,179 @@
+
+  # Pages
+  if not db.tryExec(pagesTable):
+    info("Database: pages table already exists")
```
Nice! Much prettier :+1:

ThomasTJdev commented on nimwcpkg/resources/administration/help.nim:
```diff
@@ -1,16 +1,19 @@
+  ∿∽∿∽∿∽∿∽∿∽∿∽∿∽∿∽∿∽∿∽∿∽∿∽∿∽∿∽∿∽∿∽∿∽∿∽∿∽∿∽∿∽∿∽∿∽∿∽∿∽∿∽∿∽∿∽
 
-    Nim Website Creator
+    Nim Website Creator 👑 https://NimWC.org
```
I guess it's an icon? For me it's just a square.. :disappointed:

ThomasTJdev replied:
Ahh it's a crown :smiley: 
👑 👑 👑 👑 👑 👑 👑 👑 👑 👑 👑 👑

ThomasTJdev commented on nimwcpkg/resources/email/email_connection.nim:
```diff
@@ -1,85 +1,79 @@
 
   try:
+    await client.connect(smtpAddress, Port(parseInt(smtpPort)))
+    await client.auth(smtpUser, smtpPassword)
```
Really good idea to include connect and auth in the try!! :+1:

ThomasTJdev commented on nimwcpkg/resources/email/email_connection.nim:
```diff
@@ -1,85 +1,79 @@
-  ## Send the email through smtp
-  ## Clean admin mailing
-
+  ## Send the email through smtp. Clean admin mailing.
```
Maybe?
```nim
## Send email only to admin
```

ThomasTJdev commented on nimwcpkg/resources/web/routes.nim:
```diff
@@ -21,41 +27,36 @@ routes:
     resp genMain(c, "<h3 style=\"text-align: center; color: red; margin-top: 100px;\">" & decodeUrl(@"errorMsg") & "</h3>")
 
 
+# Plugins #####################################################################
```
See comment about macros

ThomasTJdev commented on nimwcpkg/resources/web/routes.nim:
```diff
@@ -185,24 +177,21 @@ routes:
-  #[
-
-      Settings
+# Settings ####################################################################
```
See comment about macros

ThomasTJdev commented on nimwcpkg/resources/web/routes.nim:
```diff
@@ -305,48 +297,44 @@ routes:
 
 
-  #[
+# Files #######################################################################
```
See comment about macros

ThomasTJdev commented on nimwcpkg/resources/web/routes.nim:
```diff
@@ -442,14 +425,9 @@ routes:
       resp("Error: File not found")
 
 
+# Users #######################################################################
```
See comment about macros

ThomasTJdev commented on nimwcpkg/resources/web/routes.nim:
```diff
@@ -595,25 +572,17 @@ routes:
 
     resp("Error: Something went wrong")
 
+# Blog ########################################################################
```
See comment about macros

ThomasTJdev commented on nimwcpkg/resources/web/routes.nim:
```diff
@@ -680,19 +646,13 @@ routes:
     resp genPageBlog(c, blogid)
 
 
+# Pages #######################################################################
```
See comment about macros

ThomasTJdev commented on nimwcpkg/resources/web/routes.nim:
```diff
@@ -757,11 +714,9 @@ routes:
 
 
-  #[
+# Sitemap #####################################################################
```
See comment about macros

juancarlospaco commented on nimwcpkg/nimwc_main.nim:
```diff
@@ -64,21 +71,22 @@ macro configExists(): untyped =
-#[
-      Macros
-__________________________________________________]#
+# Macros ######################################################################
```
3️⃣

juancarlospaco commented:
ormin is basically SQL Pseudo-code, so if you understand SQL, you kinda already know ormin.

Later I will add instructions for Docker and Vagrant.

All feedback is implemented.

juancarlospaco commented:
Whats the difference between `defined(dev)` and `not defined(release) and not defined(demo)`?,

because `dev` has very few real uses on the code, most of uses are like:

```nim
when defined(dev):
  echo something
```

What if we drop `dev` but keep `devemailon` :grey_question:

ThomasTJdev commented:
> Literally copy & pasted code without changing it from `nimwc` to `datetime2human` since we have very similar code,
> migrated from NimWC to clean out code, benefits both projects, and added you as a repo admin there too.

Nice, good choice. I like the names `now2human()` and `datetime2human()` :smile:

juancarlospaco commented:
I hit a critical problem with `ormin` when it comes to full usage of it and the lack of documentation makes it even worse, 
it seems to tied to Karax and does not like mixing with manual raw queries, so that alone is a No-Go for the project.

Then I improved Gatabase adding SQLite via a compile time parameter and other improvements, 
it may not have all features that ormin supposedly it have but things actually work, 
and is ours to build improvements on top of it while keeping it friendly to raw queries and JS agnostic,
I searched all GitHub/Nimble and theres no ORMs at all, I will be using it instead of ormin, 
benefits both projects, and added you as a repo admin there too.

juancarlospaco commented:
![totp](https://user-images.githubusercontent.com/1189414/51802874-06889d80-222d-11e9-8fde-ed5017f7f524.jpg)

I will add a link to FreeOTP there too. On Demo mode dont really checks the number.

juancarlospaco commented:
- 2FA is a field on User table.
- 2FA Secret is Base32 as best practice according to standard.
- Added Terms Of Service (TOS) page, pointing to a static HTML TOS Template.
- Added Bulma styles to Logs page.
- Added Bulma styles to Server Info page.

ThomasTJdev commented on nimwcpkg/resources/administration/create_standarddata.nim:
```diff
@@ -1,4 +1,4 @@
-# Copyright 2018 - Thomas T. Jarløv
+# Copyright 2019 - Thomas T. Jarløv
```
*Comment:* I have no need for my name at the top of every file, and I do think **our** license GPL and copyright could just be defined at the main file, which scales down and includes all of the other files.

ThomasTJdev commented:
> ... firejail
> .. People seems to love it:
> ...

And I do to! THIS IS AWESOME! I wasn't aware of firejail, but I'm gonna dig into it now.

juancarlospaco commented:
![nimwc-temp0](https://user-images.githubusercontent.com/1189414/52220062-bd63c980-287c-11e9-9cfc-b716a407b8b3.png)

juancarlospaco commented:
![nimwc_design](https://user-images.githubusercontent.com/1189414/52374791-f0ed5200-2a3c-11e9-969e-7fdc99eef0c6.jpg)

![nimwc-design](https://user-images.githubusercontent.com/1189414/52374807-f64a9c80-2a3c-11e9-8687-c870c71ee7d7.png)

- `-d:noUnsplash` to Disable.

juancarlospaco commented:
Postgres
```sql
(extract(epoch from now()))
```

Gives:
```
Error: unhandled exception: ERROR:  column "creation" is of type timestamp but default expression is of type double precision
```

I changed it to `now()`, but can be improved, it seems it accepts ISO formatted but not literal numbers?.

juancarlospaco commented:
Ready to test it out @ThomasTJdev :tada: :boom: :fireworks:

ThomasTJdev commented:
> Ready to test it out @ThomasTJdev    To try all new stuff...
> I tested this on another PC and worked ok.

Great! I'll try add a checklist, please pitch in, so we are testing everything.

> Do you prefer `-d:demo` to automatically force `-d:sqlite -d:noWebp -d:noUnplash`

I don't think so. I'll evaluate during the test.

ThomasTJdev commented:
- [ ] General UI check
- [ ] Files - UI
- [ ] Files - Folder (from image)
- [ ] Files - Count (#45)
- [ ] Plugin - UI
- [ ] Plugin - Enable/Disable route
- [ ] Docket
- [ ] Firejail
- [ ] WebP
- [ ] Unsplash
- [ ] Postgres
- [ ] DevOps folder
- [ ] Datetime2human integration
- [ ] Online config
- [ ] Admin panel
- [ ] Honeypot
- [ ] 2FA
- [ ] Password generater (random)
- [ ] Logs - UI
- [ ] Nimble install on a clean system
- [ ] Nim latest devel
- [ ] Check changelog

**Other**
- [ ] Error using Summernote with Bulma (#37)
- [ ] Nim's symlink proc cant overwrite existing symlink (#36)
- [ ] Version bump to 5

juancarlospaco commented:
I did not meant to Fix #36 , #37 , #45 nor Bump to 5, but I think is good to have small iterations, 
and come back to `master` before Diff gets uncontrollably huge and hard to review.

`Dockerfile` is meant as a template, something you copy over and modify, not something that will run as-is.

ThomasTJdev commented:
This is a really nice PR with so many great features and improvements!! Great work @juancarlospaco !

_______


# 2FA
Would it be okay with you, if I make these changes?
1) I would like to make it optional to enable 2fa.
2) The 2fa random-key should be generated and shown to the user. The user can then use a 2fa-software/app to save it.
3) The user needs to actively enable 2fa from the profile page.

To do: We should first check pwd and email, if they match we should ask for 2FA. Or we can keep the current layout to force the user to think about 2FA on every login.


## nimwc2fa.nim
Can't get the file to work:
```nim
os.nim(1385)             paramStr
Error: unhandled exception: invalid index [IndexError]
Error: execution of an external program failed: './nimwc2fa '
```


# Unsplash
I couldn't get it to work.. where should the pictures change? How to use it? My main folder is getting the pics.
Would it be an idea to move it to a plugin?


# TOS
Could be expandede? Like https://cxmanager.dk/eula ?
The font didn't work for me. I changed it to font-family sans-serif.
Shouldn't we make it editable? Like the config file. Then the user can specify the details to comply with GDPR.


# -d:noFields
Purpose?


# History table
Purpose?


# Postgres "WITHOUT OIDS"
Is this needed? https://www.postgresql.org/docs/9.6/datatype-oid.html
"Object identifiers (OIDs) are used internally by PostgreSQL as primary keys for various system tables. OIDs are not added to user-created tables, unless WITH OIDS is specified when the table is created, or the default_with_oids configuration variable is enabled. "


# Webp
The quality is really low - we should make it optional, where the user can choose to use webp.


# nim.cfg
I have added -d:noFirejail and -d:noUnsplash into the cfg file. Firejail requires foreign dependency and I'm not sure of unsplash.
I'm okay by including both, but I think they should be disabled as standard, and the user could then enable them from the UI and NimWC could then recompile to enable them?


# To do:
`files.tmpl` should use the files-table instead of walkdir on `files/privat-public`. We can then also show how many times, a file has been downloaded.

Split readme features into categories for better overview

Update NimWC.org with information about all the new features :)

juancarlospaco commented on nimwc.nim:
```diff
@@ -25,7 +25,7 @@ const
     ☑️ Using params: """  ## Message to show when started Compiling.
 
-  compile_ok_msg =  """👑 Nim Website Creator Compiled Ok! 👑
+  compile_ok_msg =  """ Nim Website Creator Compiled Ok!
```
Oh my bad, what Terminal this wont work?,
Ive tried KDE, Gnome, LXQt, Server without X and Mac and works Ok. 
Only Windows wont work, but we dont support that.
I wonder if its a bug of some Terminal apps :thinking: 
Anyways wont worth wasting time on this. :+1:

ThomasTJdev replied:
Nooo! I really liked the crowns!! I have tried XTerm and Terminator - both with Xfce4. Maybe it's a font-package I need?

juancarlospaco replied:
Then is a Distro Packaging Bug?, Terminal should include the font to display Unicode on the package dependencies IMHO.

XFCE dont have a Release since 2015 https://en.wikipedia.org/wiki/Comparison_of_X_Window_System_desktop_environments#Overview
(individual apps received patches).

ThomasTJdev replied:
:open_mouth: .. well, it looks like I'm a little outdated.. Let's add the crown again, sorry.

ThomasTJdev commented:
![2fa](https://user-images.githubusercontent.com/1888696/52896873-efa2ee80-31cd-11e9-9c25-d5fcc8a98458.png)
![2faerror](https://user-images.githubusercontent.com/1888696/52896875-f3cf0c00-31cd-11e9-9b79-d2bd8d47d4c9.png)
![2fasuccess](https://user-images.githubusercontent.com/1888696/52896876-f6316600-31cd-11e9-8d77-c5a9b72f6c3f.png)
![2faenabled](https://user-images.githubusercontent.com/1888696/52896877-f893c000-31cd-11e9-8d48-4f64ac29e299.png)

ThomasTJdev commented:
I would like to delete `nimwc2fa.nim` after df183369e81c40ad9853ab2cb193de2836356b30.

Ok?

ThomasTJdev commented:
I think we should go for a version 5 :smiley: . What do you think? c03f7eea5884923b3f5d9e957685c313f5ecabe2

ThomasTJdev commented:
Gatabase is only used when connecting to the DB. If you don't use in other places, I suggest we remove it for now to strip down dependencies.

ThomasTJdev commented:
I have tried to split the README into sections. I think it needs a double check 65b9202458be9227cd0a64126925d7895411f60f

juancarlospaco commented on nimwcpkg/resources/web/routes.nim:
```diff
@@ -634,6 +634,48 @@ routes:
+
+    try:
+      if $newTotp(@"twofakey").now() == @"testcode":
+        resp("Success, the code mathed")
```
Typo `mathed`

ThomasTJdev commented on nimwcpkg/resources/administration/create_adminuser.nim:
```diff
@@ -23,31 +23,32 @@ proc createAdminUser*(db: DbConn, args: seq[string]) =
-    let
-      salt = makeSalt()
-      password = makePassword("test", salt)
+    const salt = "0".repeat(128)  # Weak Salt for Test user only.
```
Why use a weak salt?

juancarlospaco replied:
Because is a `const` and the password is literally written on the page, 
no point to use a strong Salt if is for Demostration purposes only.

ThomasTJdev commented on nimwcpkg/tmpl/user.tmpl:
```diff
@@ -269,9 +269,23 @@
+                </details>
                 # proc twoFaKey(): string =
-                #   let items = toSeq('A'..'Z')
+                #   const items = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z']
```
Good thought creating a constant :+1:

ThomasTJdev commented on nimwcpkg/tmpl/user.tmpl:
```diff
@@ -269,9 +269,23 @@
+                    <li>Interval:  30  </li>
+                    <li>Phone time must be correct.</li>
+                    <li>Bogus codes like 000000 or 999999 or similar are ignored.</li>
+                    <li>Phone must be configured to sync time with cellular network.</li>
```
Isn't `Phone must be configured to sync time with cellular network` overkill? The phone could be in flight mode?

juancarlospaco replied:
Phone on flight mode still syncs time AFAIK, but I will remove it then.

ThomasTJdev replied:
Well I didn't know that - up to you :)

ThomasTJdev commented on nimwcpkg/tmpl/serverinfo.tmpl:
```diff
@@ -17,7 +17,7 @@
 #  else:
 #    let hostn = getHostname()
-#    let uptim = staticExec("uptime --pretty").strip
+#    let uptim = execCmdEx("uptime --pretty").output.strip
```
Good catch!

ThomasTJdev commented on nimwcpkg/resources/administration/createdb.nim:
```diff
@@ -28,7 +28,7 @@ const
       password   varchar(300)  not null,
       twofa      varchar(60),
-      email      varchar(254)  not null,
+      email      varchar(254)  not null           unique,
```
:+1:

ThomasTJdev commented on nimwcpkg/resources/web/routes.nim:
```diff
@@ -511,7 +511,7 @@ routes:
       when defined(webp):
         if path.endsWith(".png") or path.endsWith(".jpg") or path.endsWith(".jpeg"):
-          discard cwebp(path, path, "text", quality=9)
+          discard cwebp(path, path, "drawing", quality=25)  # This sets quality of WEBP
```
We could let the user set the quality in the config file and just have 25 as standard? Some might prefer higher or lower. We could even add a small option when uploading pictures, that the user can change the quality from the browser.

juancarlospaco replied:
I was thinking the same. I got more ideas for future roadmap.

ThomasTJdev commented on nimwcpkg/tmpl/user.tmpl:
```diff
@@ -192,7 +192,7 @@
       <td> Mail unconfirmed </td>
     #   end if
-    <td> ${ user[6].len > 6 } </td>
+    <td> ${ if user[3] == "Admin": toUpperAscii($(user[6].len > 6)) else: "???" } </td>
```
Why use `???` instead of `FALSE`?

ThomasTJdev replied:
I made a mistake! This will look wrong, if a non-Admin user sees it - then it would say `FALSE` even it its true.. should we hide the field from non-Admin users? `when c.rank == Admin`

ThomasTJdev commented on nimwcpkg/nimwc_main.nim:
```diff
@@ -343,8 +343,8 @@ proc login(c: var TData, email, pass, totpRaw: string): tuple[b: bool, s: string
-        if totp == 000000 or totp == 999999 or totp == 123456 or totp == 654321 and not defined(demo):
-          return (false, "2 Factor Authentication Number must not be 000000 or 999999 or 123456")
+        if totp in [000000, 111111, 222222, 333333, 444444, 555555, 666666, 777777, 888888, 999999]:
+          return (false, "2 Factor Authentication Number must not be 000000 or 999999 or similar")
```
Could be `2 Factor Authentication Number must not be 6 identical numbers, e.g. 000000 or 999999`.

Up to you.

juancarlospaco commented:
I would like to Merge this for the next month or before, like 1 march or near, 325 Commits already.
Looks good to me, what do you think @ThomasTJdev :grey_question:

ThomasTJdev commented:
> I would like to Merge this for the next month or before, like 1 march or near, 325 Commits already.
> Looks good to me, what do you think @ThomasTJdev

Agree. I would like to do some more test, but we will merge before monday the 4th of March - hopefully before?

juancarlospaco commented:
Just for historical reference:

DropBox added Unsplash: https://medium.com/unsplash/unsplash-for-dropbox-paper-d7c0c425d2d0
Sketch, Adobe, Trello, Google, Medium, Square, Semplice, Typeform, Weebly, PicsArt, Ghost, CodePen, Unfold, Product Hunt, Anchor, InVision, Unbounce, FiftyThree already had Unsplash.

Docker only is Insecure: https://snyk.io/blog/top-ten-most-popular-docker-images-each-contain-at-least-30-vulnerabilities

ThomasTJdev commented:
> DropBox added Unsplash: https://medium.com/unsplash/unsplash-for-dropbox-paper-d7c0c425d2d0
> Sketch, Adobe, Trello, Google, Medium, Square, Semplice, Typeform, Weebly, PicsArt, Ghost, CodePen, Unfold, Product Hunt, Anchor, InVision, Unbounce, FiftyThree already had Unsplash.

Unsplash is currently not active as standard (not present in `nimwc.nim.cfg`). When you create a website - blog, portfolio etc. don't you think, that the user would like to choose static background image? I'm all in for having it as a part of NimWC, but I would like the user to manually turn it on.

____________________

I would like to implement these 2 things before release - ok?
- a view-counter on blogpost (like the one on file-downloads) 
- a "fake" publishing date for blogpost. You can write a date, which is shown in the blogoverview. Sometimes you might want to create a blogpost in private mode and publish it later, but the end-user is not supposed to know when you created it.

juancarlospaco commented:
I removed Unsplash and Gatabase.

For this Pull requests I go with vanilla Postgres but we may [implement Norm later on](https://github.com/ThomasTJdev/nim_websitecreator/issues/55),
and for Unsplash a better UX is needed so maybe it come back someday. I think is important to come back to master.

The features you mention sounds kind of cool!, 
but I dont know how to implement them properly from that short description, 
but no problem just push it and ship it!. I am just super happy with having Firejails in. :smile:

ThomasTJdev commented:
> * You can choose how much CPU & RAM NimWC can use.
> 
> https://gist.github.com/juancarlospaco/9ef1b803ef0cf6f71cde2dbec6839116#file-example-txt

Nice! Will you write a small `How to config firejail` for the `README`? Or can set the default settings, so it'll work on most systems (RPi's, EC2 tier, etc.)?

ThomasTJdev commented:
Good update on firejail in the readme :+1:

ThomasTJdev commented:
Question: What should our users standard DB be? SQLite or Postgres?

We currently have `-d:sqlite` in the nimwc.nim.cfg file, but inside the code we use `when defined(sqlite)`. This is double negative, and the user needs to remove the line from nimwc.nim.cfg to enable postgres - there's no other way to force postgres. If sqlite is our primary DB, we should use `when defined(postgres)` and remove `-d:sqlite` from nimwc.nim.cfg.

I would prefer sqlite as standard. It makes it easier for a user to try NimWC without setting up a Postgres DB.

What do you think?

juancarlospaco commented:
@ThomasTJdev I agree :+1:

ThomasTJdev commented on nimwcpkg/tmpl/utils.tmpl:
```diff
@@ -182,7 +182,7 @@
     <div class="control">
       <label class="label">Publication date (visible on the blog overview)</label>
-      <input class="input" type="text" name="pubdate" value="${pubDate}" dir="auto">
+      <input class="input" type="date" name="pubdate" value="${pubDate}" min="${pubDate}" dir="auto" pattern="[0-9]{4}-[0-9]{2}-[0-9]{2}" required >
```
This shows the date as MM/DD/YYYY - instead of YYYY-MM-DD or DD/MM/YYYY.

juancarlospaco replied:
Thats Cosmetic, the Date is Auto Locale aware.
Over the wire is just `YYYY-MM-DD`

ThomasTJdev commented:
@juancarlospaco I'm ready for merging :smiley: . You just say go, when you are ready!

juancarlospaco commented:
GO :grey_exclamation:

juancarlospaco commented:
@ThomasTJdev Need to Re-Deploy the Demo.
:fireworks: :fireworks: :fireworks: :fireworks: :fireworks: :fireworks: :fireworks: :fireworks:

ThomasTJdev commented:
I'm having 2 problems :disappointed: 

## Nimble file
Should we remove this line? It is a bin-file, so Nimble should compile it itself.
```nim
exec "nim c -r -d:release -d:ssl nimwc.nim"
```

## Firejail
When running I get this:
```nim
ubuntu@development:~/tmp/nim_websitecreator$ nimwc
2019-03-01T21:54:21+00:00: Nim Website Creator: Launcher starting.
Error: invalid --keep-dev-shm command line option
2019-03-01T21:54:23+00:00: Restarting in 1 second.
Error: invalid --keep-dev-shm command line option
2019-03-01T21:54:26+00:00: Restarting in 1 second.
^CError: unhandled exception: No such process [OSError]
```

juancarlospaco commented:
Remove the line if you want, you are right.

For the Demo you can compile without Firejail.
Nimble should copy that file anyways, you are right.

Are you sure you have the latest Firejail module?, you ran `nimble refresh` ?.

I updated both Firejail and WebP during this Pull Request, so reinstall them to update them.

ThomasTJdev commented:
This is a clean test install the server. I can compile without firejail.

I have the newest firejail-nim module from nimble.

Firejail version:
```
ubuntu@development:~/.nimble/pkgs/nimwc-5.0.0$ firejail --version
firejail version 0.9.52

Compile time support:
	- AppArmor support is enabled
	- AppImage support is enabled
	- bind support is enabled
	- chroot support is enabled
	- file and directory whitelisting support is enabled
	- file transfer support is enabled
	- git install support is disabled
	- networking support is enabled
	- overlayfs support is enabled
	- private-home support is enabled
	- seccomp-bpf support is enabled
	- user namespace support is enabled
	- X11 sandboxing support is enabled
```

I'm using the raw config.cfg - no changes has been made.

juancarlospaco commented:
`0.9.58.2` is latest release of Firejail https://firejail.wordpress.com/download-2/

If you want leave the Demo without Firejail, whatever you like no problem.

ThomasTJdev commented:
> `0.9.58.2` is latest release of Firejail https://firejail.wordpress.com/download-2/
> 
> If you want leave the Demo without Firejail, whatever you like no problem.

Demo without firejail is ok. 

Is v`0.9.58.2` required? On ubuntu I can't get higher than *.52 on standard library:
```
ubuntu@development:~/.nimble/pkgs/nimwc-5.0.0$ apt search firejail
Sorting... Done
Full Text Search... Done
firejail/bionic,now 0.9.52-2 amd64 [installed]
  sandbox to restrict the application environment

firejail-profiles/bionic,now 0.9.52-2 all [installed,automatic]
  profiles for the firejail application sandbox

firetools/bionic 0.9.50-1 amd64
  Qt frontend for the Firejail application sandbox
```

juancarlospaco commented:
They have the `*.DEB` packages here: https://sourceforge.net/projects/firejail/files/firejail/ (is Pure C).

ThomasTJdev commented:
> They have the `*.DEB` packages here: https://sourceforge.net/projects/firejail/files/firejail/ (is Pure C).
> 
> `--keep-dev-shm` is Not being used, if you like just remove the line from `nim-firejail`.
> 
> But being kinda security package I dont feel like supporting legacy or deprecated too much lol.
> 
> Whatsoever is Distros fault not our fault.

I like the security approach, but it could make it hard to support AWS Ubuntu images or Raspberry Pi's. RPi has a backport version 9.44. 

Can you compile on a RPi? I'm having trouble (with and without firejail).
```nim
nimwc_main.nim(582, 15) template/generic instantiation of `generateRoutes` from here
/home/pi/nim/Nim/lib/core/macros.nim(496, 1) template/generic instantiation of `routes` from here
../../jester-0.4.1/jester.nim(1269, 35) template/generic instantiation of `async` from here
resources/email/email_registration.nim(48, 6) Warning: 'sendEmailActivationManual' is not GC-safe as it calls 'sendEmailActivationManual_continue' [GcUnsafe2]
nimwc_main.nim(582, 15) template/generic instantiation of `generateRoutes` from here
/home/pi/nim/Nim/lib/core/macros.nim(496, 1) template/generic instantiation of `routes` from here
../../jester-0.4.1/jester.nim(1269, 35) template/generic instantiation of `async` from here
/home/pi/nim/Nim/lib/pure/asyncmacro.nim(267, 31) Warning: 'matchIter' is not GC-safe as it calls 'sendEmailActivationManual' [GcUnsafe2]
/home/pi/nim/Nim/lib/core/macros.nim(1214, 49) Error: conversion from int64 to int is invalid
 Compile Error

  ⚠️ Compile-time or Configuration or Plugin error occurred.
  ➡️ You can check your source code with: nim check YourFile.nim
  ➡️ Check the Configuration of NimWC and its Plugins.
  ➡️ Remove new Plugins, restore previous Configuration.
   Check that you have the latest Version. Check the Documentation.
```

ThomasTJdev commented:
> They have the `*.DEB` packages here: https://sourceforge.net/projects/firejail/files/firejail/ (is Pure C).
> 
> `--keep-dev-shm` is Not being used, if you like just remove the line from `nim-firejail`.
> 
> But being kinda security package I dont feel like supporting legacy or deprecated too much lol.
> 
> Whatsoever is Distros fault not our fault.

www.nimwc.org is up and running with firejail :smiley: *(I need to perform some upgrades, so there might be a little downtime next hour)*

If `--keep-dev-shm`  isn't used, could it then be comment out. And if it's going to be used in the future, it can be reimplemented?

juancarlospaco commented:
I just bought this to test Nimwc on ARM, but did not have time to install the OS yet. 

![dscn0013](https://user-images.githubusercontent.com/1189414/53681681-a5406800-3ccb-11e9-86be-6cd5888d17a2.JPG)

Are you sure the OS is 64Bit?. 
The versions available for ARM should be the same versions than desktop.
Ok about the `--keep-dev-shm`.

ThomasTJdev commented:
> Can you compile on a RPi? I'm having trouble (with and without firejail).
>

I have tried compiling from source, but with no luck. Have you had any luck upgrading firejail on a RPi? If we can't support RPi out of the box, we should:
1. Inform about it in the readme and do a os-check to avoid compiling with firejail
2. Check local firejail version, and only compile with firejail if version is supported
3. Update nim-firejail to only use "old" functions when it's an old firejail version

juancarlospaco commented:
Have you tried ArchLinux on it?, maybe is Distro-specific Bug.

ThomasTJdev commented:
> Have you tried ArchLinux on it?, maybe is Distro-specific Bug.

It's running standard [Raspbian](https://www.raspberrypi.org/downloads/). But I'm running Jessie, and the current stable is Stretch. I'm upgrading now to retest since Strecth has [9.58](https://packages.debian.org/search?suite=default&section=all&arch=any&searchon=names&keywords=firejail).

ThomasTJdev commented:
> > Have you tried ArchLinux on it?, maybe is Distro-specific Bug.
> 
> It's running standard [Raspbian](https://www.raspberrypi.org/downloads/). But I'm running Jessie, and the current stable is Stretch. I'm upgrading now to retest since Strecth has [9.58](https://packages.debian.org/search?suite=default&section=all&arch=any&searchon=names&keywords=firejail).

Stretch has firejail `0.9.58` - so firejail is fine on a RPi :+1: . But I'm still getting the conversion error from `macros.nim` with both `0.19.9` and `0.19.4`.
